### PR TITLE
Refactor Data Pipeline for Robotics Example

### PR DESCRIPTION
**Pull request by mknitt**
This builds on and modifies #311 

- Includes training, prediction, model saving and loading functionalities
- integrates with flowcean’s Model/SupervisedLearner API
- Adds logic for train/test split and concatenation across multiple datasets.
- Switched to using the new ImageBasedLightningLearner for training.
- Introduced offline evaluation with multiple metrics (Accuracy, ClassificationReport, FBetaScore, PrecisionScore, Recall) using evaluate_offline from flowcean.
- Model saving now stores the LightningModule state dict.
- Updated config.yaml to set the default rosbag.path to recordings/test_folder.
- Added /test_folder to .gitignore in the recordings directory.
- Added a test model file (models/test_folder.pt).

### CODE REVIEW

File: examples/robot_localization_failure/run.py [hunk 1]
```diff
@@ -1,103 +1,311 @@
...
+                ground_truth="/momo/pose",
+                estimation="/amcl_pose",
+                position_threshold=config.localization.position_threshold,
+                heading_threshold=config.localization.heading_threshold,
+            )
+        )
+        try:
+            data.observe().sink_parquet(out_path)
+            msg = f"Processed data saved to {out_path}"
+            logger.info(msg)
```
[the remark this answers is not in the document]
Comment (schmidma) [reply examples/robot_localization_failure/run.py:138]: applies to many future occurrences as well...

File: examples/robot_localization_failure/run.py [hunk 2]
```diff
@@ -1,103 +1,311 @@
...
+            logger.exception(msg)
+            continue
+
+    # Combinedatasets
+    train_data = pl.concat([train.data for train in train_datasets])
+    test_data = pl.concat([test.data for test in test_datasets])
+
+    # Debug: Print columns to verify structure
+    logger.info(f"Train data columns: {train_data.columns}")
+    logger.info(f"Test data columns: {test_data.columns}")
```
Comment (schmidma) [inline examples/robot_localization_failure/run.py:235]: combine into a single message

File: examples/robot_localization_failure/run.py [hunk 3]
```diff
@@ -1,103 +1,311 @@
...
+            )
+            test_datasets.append(test_images)
+        except Exception as e:
+            msg = f"Error loading {parquet_path}: {e}"
+            logger.exception(msg)
+            continue
+
+    # Combinedatasets
+    train_data = pl.concat([train.data for train in train_datasets])
+    test_data = pl.concat([test.data for test in test_datasets])
```
Comment (schmidma) [inline examples/robot_localization_failure/run.py:231]: This sounds weird... you are first building a `FeatureImagesData` and then accessing the raw polars data again and rebuilding a new dataset using the factory below

File: examples/robot_localization_failure/custom_learners/image_based_lightning_learner.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+from typing_extensions import override
+
+from flowcean.core import Model, SupervisedLearner
+
+DatasetFactory = Callable[[pl.DataFrame], Dataset]
+
+logger = logging.getLogger(__name__)
+
+
+class ImageBasedLightningLearner(SupervisedLearner):
```
Comment (schmidma) [inline examples/robot_localization_failure/custom_learners/image_based_lightning_learner.py:25]: I think this can be generalized to the default lightning learner in flowcean, and removed here.

File: examples/robot_localization_failure/run.py [hunk 5]
```diff
@@ -1,103 +1,311 @@
...
-    | Drop("/scan", "/particle_cloud", "/momo/pose", "/amcl_pose")
-    # detect experiment slice points based on delocalization events
-    | DetectDelocalizations("/delocalizations", name="slice_points")
-    | Drop("/delocalizations")
-    | SliceTimeSeries(
-        time_series="measurements",
-        slice_points="slice_points",
+    model = learner.learn(
+        inputs=train_data.drop(["is_delocalized"]),
+        outputs=train_data.select(["is_delocalized"]),
```
Comment (schmidma) [inline examples/robot_localization_failure/run.py:258]: we should be able to use the flowcean strategies here. `learn_offline` for example?

File: examples/robot_localization_failure/run.py [hunk 6]
```diff
@@ -1,103 +1,311 @@
...
+
+def get_rosbag_paths(
+    config: DictConfig | ListConfig,
+) -> tuple[Path, list[Path]]:
+    rosbag_dir = Path(config.rosbag.path)
+
+    if not rosbag_dir.is_dir():
+        msg = f"Specified ROS2 bag directory {rosbag_dir} does not exist."
+        logger.error(msg)
+        sys.exit(1)
```
Comment (schmidma) [inline examples/robot_localization_failure/run.py:292]: raise an exception instead of log and exit

File: examples/robot_localization_failure/config.yaml [hunk 7]
```diff
@@ -1,9 +1,16 @@
 rosbag:
-  path: ???
+  path: recordings/test_folder
```
Comment (schmidma) [inline examples/robot_localization_failure/config.yaml:2]: I'd suggest specifying a list of rosbag diretories to make the experiment more reproducible

File: examples/robot_localization_failure/run.py [hunk 8]
```diff
@@ -1,103 +1,311 @@
...
-)
+    print(report)
+    # Save the model
+    out_path = f"models/{rosbag_dir.name}.pt"
+    msg = f"Saving model to {out_path}"
+    logger.info(msg)
+    torch.save(model.module.state_dict(), out_path)
+
+
+def get_rosbag_paths(
```
Comment (schmidma) [inline examples/robot_localization_failure/run.py:284]: if you specify the rosbag paths in the `config.yaml` as a list, you can probably remove the entire function.

File: examples/robot_localization_failure/custom_learners/image_based_lightning_learner.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    @override
+    def save_state(self) -> dict[str, Any]:
+        """Save the model state to a dictionary.
+
+        Returns:
+            A dictionary containing the model state.
+        """
+        model_bytes = BytesIO()
+        torch.save(self.module.state_dict(), model_bytes)
+        model_bytes.seek(0)
```
[the remark this answers is not in the document]
Comment (mknitt) [reply examples/robot_localization_failure/custom_learners/image_based_lightning_learner.py:183]: https://docs.pytorch.org/tutorials/beginner/saving_loading_models.html#saving-loading-model-for-inference
